samples: crypto: Added KMU sample for nRF54L (#24802)
nrfconnect/sdk-nrf: samples/crypto/kmu_usage_nrf54l/README.rst

#. Removal of the keys from RAM.
#. Using both pre-provisioned and generated keys/key pairs to encrypt/descrypt
   or sign/verify a message or its hash.
#. Cleanup.
   All persistent keys generated by this sample will be removed from the KMU, while
   all pre-provisioned keys remain in KMU between resets.

.. note::
   The read-only type of persistent keys cannot be destroyed with the ``psa_destroy_key`` function.
   The ``PSA_KEY_PERSISTENCE_READ_ONLY`` macro is used for read-only keys.
   The key ID of a read-only key is writable again after a full erase of the device memory.
   Use the ``west -v flash --erase`` command for the full erase.

.. note::
   Builds without TF-M and all nRF54L15 builds use the :ref:`hardware unique key (HUK) <lib_hw_unique_key>` to encrypt the key before storing it.

Building and running
********************

.. |sample path| replace:: :file:`samples/crypto/kmu_usage_nrf54l`

.. include:: /includes/build_and_run_ns.txt

Keys generation and provisioning
==================================================

Prior to program the device the following keys/key pairs must be generated:

1. ed25519:

>    Refer to the `Managing keys <https://docs.nordicsemi.com/bundle/ncs-latest/page/mcuboot/imgtool.html#managing_keys>`_ section of the MCUBoot documentation.

Reviewer: Add a link target to the /doc/nrf/links.txt file and use it as advised in https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/dev_model_and_contributions/documentation/styleguide.html#linking

>    It describes the way to use `imgtool.py` to generate an ed25519 key pair.

Reviewer: ```suggestion
   It describes how to use the :file:`imgtool.py` script to generate an ed25519 key pair.
```

   The following command was used for this sample:

   .. code-block:: console

      python3 imgtool.py keygen -k my_ed25519_priv_key.pem -t ed25519

#. AES-128.

.. note::
   Sample keys have already been generated for this sample.

>    These can be found in `keys/sources <keys/sources>`_ directory.

Reviewer: ```suggestion
   These are located in the :file:`keys/sources` directory.
```

>    Please do NOT use them for production.

Reviewer: ```suggestion
   Do not use them for production.
```

> After keys have been generated, they must be put to the apecial JSON file to be further used by `nRF Util` for key provisioning.

Reviewer: ```suggestion
After the keys have been generated, put them to the special JSON file to be further used by the `nRF Util development tool`_ for key provisioning.
```

> See instructions for it in `PSA Key Attributes generator script <https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/scripts/generate_psa_key_attributes/generate_psa_key_attributes.html#generate-psa-key-attributes-script>`_ documentation.

Reviewer: ```suggestion
See instructions for it in :ref:`generate_psa_key_attributes_script`.
```

.. note::

>    Sample `keys.json <keys/keys.json>`_ file have already been prepared and can be used for provisioning.

Reviewer: ```suggestion
   The :file:`keys.json` file has already been prepared and you can use it for provisioning.
```

>    The following commands were used to generate the JSON file:\

Reviewer: ```suggestion
   The following commands were used to generate the :file:`JSON` file:
```

   .. code-block:: console

      python3 generate_psa_key_attributes.py \
         --usage ENCRYPT_DECRYPT \
         --id 1 \
         --type AES \
         --key-bits 128 \
         --location LOCATION_CRACEN_KMU \
         --cracen-usage PROTECTED \
         --algorithm CTR \
         --persistence PERSISTENCE_DEFAULT \
         --key 0x44daa497a188ad60c7962cba7c5c7d8f \
         --file keys.json

   .. code-block:: console

      python3 generate_psa_key_attributes.py \
         --usage SIGN \
         --id 2 \
         --type ECC_KEY_PAIR_TWISTED_EDWARDS \
         --key-bits 255 \
         --location LOCATION_CRACEN_KMU \
         --cracen-usage RAW \
         --algorithm EDDSA_PURE \
         --persistence PERSISTENCE_DEFAULT \
         --key-from-file sources/my_ed25519_priv_key.pem \
         --file keys.json

The last step is key provisioning described `here <https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/app_dev/device_guides/nrf54l/kmu_provision.html>`_.

> The following command have been used for provisioning:\

Reviewer: ```suggestion
The following command have been used for provisioning:
```

.. _crypto_kmu_usage_nrf54l:

Crypto: KMU usage (nRF54L series)
#################################

.. contents::
   :local:
   :depth: 2

The KMU usage sample shows how to generate and use various persistent keys using the Platform Security Architecture (PSA) APIs.
Persistent keys are stored in the Key Management Unit (KMU) of the device and retain their value between resets.

> The implementation of the PSA ITS API is provided by the Oberon PSA Crypto, which implements the driver wrapper role

Reviewer 2: First sentence should mention the KMU as the persistent key sample also does persistent keys. The point here is to showcase how the KMU can be used to store persistent keys.

Reviewer 3: Addressed.

and determines which driver to use for each cryptographic operation.
For this sample, CRACEN driver is used.

The sample is also used to demonstrate the usage of the :ref:`pre-provisioned keys <ug_nrf54l_developing_provision_kmu_provisioning>`.

Requirements
************

The sample supports the following development kits:

.. table-from-rows:: /includes/sample_board_rows.txt
   :header: heading
   :rows: nrf54l15dk_nrf54l05_cpuapp, nrf54l15dk_nrf54l10_cpuapp, nrf54l15dk_nrf54l15_cpuapp, nrf54l15dk_nrf54l15_cpuapp_ns, nrf54lm20dk_nrf54lm20a_cpuapp, nrf54lv10dk_nrf54lv10a_cpuapp

.. include:: /includes/tfm.txt

Overview
********

> In this sample, four different keys/keypairs are used:

Reviewer 2: This sentence says four, but you only list three keys?

Reviewer 3: Addressed.

.. _crypto_kmu_usage_nrf54l:

Crypto: KMU usage (nRF54L Series)
#################################

.. contents::
   :local:
   :depth: 2

The KMU usage sample demonstrates how to generate cryptographic keys for nRF54L Series devices and securely store them in the Key Management Unit (KMU).

> The sample also demonstrates how to use the :ref:`generate_psa_key_attributes_script` script to generate the :file:`keys.json` file and then provision it to the KMU using `nRF Util`_.

Reviewer 4: ```suggestion
The sample also demonstrates how to use the :ref:`generate_psa_key_attributes_script` to generate the :file:`keys.json` file and then provision it to the KMU using `nRF Util`_.
```

     - ECDSA with secp256r1 curve
     - Generated by the sample firmware.

The sample provides examples of the keys and key pairs that are provisioned before programming the sample.
You can find these examples in the :file:`keys/sources` directory.
See the `Configuration files`_ section for more information.

.. note::
   These examples are not intended for production use.

Sample operations
=================

After you provision the keys to the KMU using :file:`keys.json` and program the sample to the device, the sample performs the following operations:

1. Initialization of the Platform Security Architecture (PSA) API.
#. Generation of a persistent AES 128-bit key and ECDSA key pair.
#. Removal of the keys from RAM.
#. Encryption and decryption using both keys (key pairs) provisioned before programming the sample and keys (key pairs) generated by the sample firmware.
   Alternatively, signing and verifying a message or its hash.
#. Cleanup.
   All persistent keys (key pairs) generated by the sample firmware are removed from the KMU, while all keys (key pairs) provisioned before programming the sample remain in the KMU between device resets.

.. note::

>    The read-only type of persistent keys cannot be destroyed with the ``psa_destroy_key`` function.

Reviewer 4: ```suggestion
   The read-only type of persistent keys cannot be destroyed with the ``psa_destroy_key()`` function.
```
